### DISCUSSION TITLE
tasmota does not link when using USE_SML_M

### DISCUSSION THREAD

**Discussion by juergiboy** [arendst/Tasmota#17847] [Q&A] [answered]
Hi,
I already have some experience with tasmota and building custom version with VSC and PIO, but now I want to build a SML binary and that doesn't work at all.

**What I have done:**
Download the actual development branch from github and open in VSC.
Uncomment the tasmota env in platformio_override.ini
I added this 
  #ifndef USE_SCRIPT
  #define USE_SCRIPT
  #endif
  #ifndef USE_SML_M
  #define USE_SML_M
  #endif
  #ifdef USE_RULES
  #undef USE_RULES
  #endif

to the user_config_override.h as described here: [https://tasmota.github.io/docs/Smart-Meter-Interface/](url)

**This is the result:**
Retrieved `.pio\build\tasmota\FrameworkArduino\umm_malloc\umm_info.c.o' from cache
Retrieved `.pio\build\tasmota\FrameworkArduino\umm_malloc\umm_integrity.c.o' from cache
Retrieved `.pio\build\tasmota\FrameworkArduino\umm_malloc\umm_local.c.o' from cache
Retrieved `.pio\build\tasmota\FrameworkArduino\umm_malloc\umm_malloc.cpp.o' from cache
Retrieved `.pio\build\tasmota\FrameworkArduino\umm_malloc\umm_poison.c.o' from cache
Retrieved `.pio\build\tasmota\libFrameworkArduino.a' from cache
Linking .pio\build\tasmota\firmware.elf
c:/users/josch/.platformio/packages/toolchain-xtensa@2.40802.200502/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: .pio\build\tasmota\lib3da\libams.a(han_Parser.cpp.o):(.text._ZN10Han_Parser10unwrapDataEPhR17DataParserContext+0x24): undefined reference to `GCMParser::GCMParser(unsigned char*, unsigned char*)'
c:/users/josch/.platformio/packages/toolchain-xtensa@2.40802.200502/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: .pio\build\tasmota\lib3da\libams.a(han_Parser.cpp.o):(.text._ZN10Han_Parser10unwrapDataEPhR17DataParserContext+0x28): undefined reference to `GCMParser::parse(unsigned char*, DataParserContext&)'
c:/users/josch/.platformio/packages/toolchain-xtensa@2.40802.200502/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: .pio\build\tasmota\lib3da\libams.a(han_Parser.cpp.o): in function `Han_Parser::unwrapData(unsigned char*, DataParserContext&)':    
han_Parser.cpp:(.text._ZN10Han_Parser10unwrapDataEPhR17DataParserContext+0x12a): undefined reference to `GCMParser::GCMParser(unsigned char*, unsigned char*)'
c:/users/josch/.platformio/packages/toolchain-xtensa@2.40802.200502/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: han_Parser.cpp:(.text._ZN10Han_Parser10unwrapDataEPhR17DataParserContext+0x13a): undefined reference to `GCMParser::parse(unsigned
char*, DataParserContext&)'
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\tasmota\firmware.elf] Error 1
============================ [FAILED] Took 64.54 seconds==================================

**Additional hint**: 
When i comment only the USE_SML_M the build is successful, the Scripting seems not to be the problem!?

I'm at a loss, can anyone help? Thanks!

**Comment by gemu2015** [comment, marked as answer]
SML decryption is now default and needs TLS
So either use
#define USE_TLS

or exclude encryption
#define NO_USE_SML_DECRYPT

**Reply by juergiboy** [reply]
That's it - it works!
Thank you!

**Reply by barbudor** [reply]
@gemu2015 @Jason2866 
If decryption is default and requires TLS, we may have to include a "fix" rule at the end of `tasmota_configurations.h` to force `USE_TLS` 

```
#if defined(USE_SCRIPT) && defined(USE_SML_M) && !defined(NO_USE_SML_DECRYPT)
#define USE_TLS
#endif
```

**Reply by Jason2866** [reply]
@barbudor It does work with enabled TLS, so your proposed change is a good Idea!

**Reply by barbudor** [reply]
done https://github.com/arendst/Tasmota/pull/17852
